### ISSUE TITLE
filesystem features check before installation

### ISSUE THREAD

**Issue by Bronek** [zfsonlinux/zfs#3234]
I was running test system with current master and yesterday I decided to try new tag 0.63-r1.3 on the same system. This didn't go too well - my root filesystem failed to mount, with message reported:

```
The pool cannot be accessed on this system because it uses the
following feature(s) not supported on this system:
com.delphix:hole_birth
com.delphix:embedded_data
```

I guess I should have spotted before/during installation that master had these new features and also that these are not available in tag 0.63-r1.3 and that's the problem:

Is there any binary or script I could/should run **before** starting different zfs version (possibly with new kernel) to ensure that the zfs modules being installed will support all features of currently mounted pools - including root filesystem (if it happens to be zfs)? This implies that this binary would be running while **old** zfs/spl modules are loaded in kernel and also that the user will be **unable** to unmount/remount any pool for the purpose of check.

**Comment by DeHackEd** [comment]
Seems like a variation of #2546 but you didn't know about it.

ZFS is backwards compatible - 0.6.4 will import pools from 0.6.3 and even ZFS-Fuse. But by using future versions you shot yourself in the foot.

**Comment by Bronek** [comment]
I think this is not the same. Issue #2546 is about creating a pool without all features enabled; this one is about helping users to avoid features "downgrade" which would prevent existing pools from being loaded by binaries being installed.

The background here is that I failed to notice that tag 0.63-r1.3 does not support all features present in master, and enabled in my root filesystem, which of course resulted in failure to load root filesystem on system restart. This was fixed with a rescue boot ISO with embedded current zfs version, but I'd rather avoid it.

I am asking what could I have done better to avoid similar trouble in the future.

**Comment by tomposmiko** [comment]
I don't fully understand your issue.
After upgrading zfs, you also had to run zpool upgrade manually:

```
# zpool status
  pool: tank
 state: ONLINE
status: Some supported features are not enabled on the pool. The pool can
	still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
	the pool may no longer be accessible by software that does not support
	the features. See zpool-features(5) for details.
  scan: scrub repaired 0 in 23h43m with 0 errors on Sat Mar 14 09:41:54 2015
```

It's a quite detailed description, what's going on.
However, perhaps it could be more accurate.

**Comment by Bronek** [comment]
@tomposmiko Nope, it's the other way around. I was not an upgrade - I wanted to test tag 0.63-r1.3 on a system created with ZFS version build from zfsonlinux/spl@03a78353 and zfsonlinux/zfs@d958324f . After the change it turned out that root filesystem cannot be mounted, because it made use of features which are not available in tag 0.63-r1.3 . You can think of it as a "downgrade", see the error message in my first post. I had to switch zfs version using rescue ISO with builtin zfs.

Ideally I should have verified that zfs version being installed supports all features of my pools before I restarted my system, but there do not seem to be good way of doing that - except perhaps for following every commit.

**Comment by tomposmiko** [comment]
You're on master branch, it's part of the game.
However, you're right, it would make life a little bit easier.

**Comment by kpande** [comment]
this would be awesome for people running ZFS root installations on remote headless servers with no IPMI or KVM access for when things go wrong.

**Comment by behlendorf** [comment]
Setting the `zpool bootfs=pool/dataset` property on the root dataset will prevent ZFS from enabling incompatible features there.  That should help but it might not be a 100% solution.

**Comment by Bronek** [comment]
I do not think that's the case. On my test system, bootfs was created by zfs version built from master few months ago, which I continued to use until recently. Then I tried to build and install zfs from tag 0.63-r1.3 which did not work, since the pool used as bootfs failed to mount with error message see above. Clearly some filesystem features used by bootfs pool are not present in tag 0.63-r1.3 (which is fine) and I was unable to foresee this before system restart (which is bad).

General case is this: build zfs version at commit Z and make a new bootfs with this version (e.g. build new test system with version Z), then for whatever reason roll back to zfs version at commit A, but then it turns out that between these two commits a filesystem feature was added. This means that rolled back zfs version will fail to mount bootfs. So far so expected, but ideally the user should be able (without the neccessite to read all commits between A and Z) to verify that the system will (or will not) be able to mount all pools **before** restarting the system with rolled back binary. For example, invoke a user mode binary which "knows" all supported features from the current build and is able to compare this list against features used by all currently mounted pools, without the necessity to replace any module in runtime. When ran before system restart, this would provide user with enough warning that, after the restart, certain pools will not be able to mount due to downgrade.
